core/src/main/java/org/apache/iceberg/PartitionStats.java

```diff
@@ -0,0 +1,249 @@
+  private long totalRecordCount;
+  private Long lastUpdatedAt; // null by default
+  private Long lastUpdatedSnapshotId; // null by default
+
+  public PartitionStats(PartitionData partition, int specId) {
+    this.partition = partition;
+    this.specId = specId;
+  }
+
+  public PartitionData partition() {
+    return partition;
+  }
+
+  public int specId() {
+    return specId;
+  }
+
+  public long dataRecordCount() {
+    return dataRecordCount;
+  }
+
+  public int dataFileCount() {
+    return dataFileCount;
+  }
+
+  public long totalDataFileSizeInBytes() {
+    return totalDataFileSizeInBytes;
+  }
+
+  public long positionDeleteRecordCount() {
+    return positionDeleteRecordCount;
+  }
+
+  public int positionDeleteFileCount() {
+    return positionDeleteFileCount;
+  }
+
+  public long equalityDeleteRecordCount() {
+    return equalityDeleteRecordCount;
+  }
+
+  public int equalityDeleteFileCount() {
+    return equalityDeleteFileCount;
+  }
+
+  public long totalRecordCount() {
+    return totalRecordCount;
+  }
+
+  public Long lastUpdatedAt() {
+    return lastUpdatedAt;
+  }
+
+  public Long lastUpdatedSnapshotId() {
+    return lastUpdatedSnapshotId;
+  }
+
+  /**
+   * Updates the partition stats from the data/delete file.
+   *
+   * @param file the ContentFile from the manifest entry.
+   * @param snapshot the snapshot corresponding to the live entry.
+   */
+  public void liveEntry(ContentFile<?> file, Snapshot snapshot) {
+    this.specId = Math.max(this.specId, file.specId());
+
+    switch (file.content()) {
+      case DATA:
+        this.dataRecordCount += file.recordCount();
+        this.dataFileCount += 1;
+        this.totalDataFileSizeInBytes += file.fileSizeInBytes();
+        break;
+      case POSITION_DELETES:
+        this.positionDeleteRecordCount += file.recordCount();
+        this.positionDeleteFileCount += 1;
+        break;
+      case EQUALITY_DELETES:
+        this.equalityDeleteRecordCount += file.recordCount();
+        this.equalityDeleteFileCount += 1;
+        break;
+      default:
+        throw new UnsupportedOperationException("Unsupported file content type: " + file.content());
+    }
+
+    if (snapshot != null) {
+      updateSnapshotInfo(snapshot.snapshotId(), snapshot.timestampMillis());
+    }
+
+    // Note: Not computing the `TOTAL_RECORD_COUNT` for now as it needs scanning the data.
+  }
+
+  /**
+   * Updates the modified time and snapshot ID for the deleted manifest entry.
+   *
+   * @param snapshot the snapshot corresponding to the deleted manifest entry.
+   */
+  public void deletedEntry(Snapshot snapshot) {
+    if (snapshot != null) {
+      updateSnapshotInfo(snapshot.snapshotId(), snapshot.timestampMillis());
+    }
+  }
+
+  /**
+   * Appends statistics from given entry to current entry.
+   *
+   * @param entry the entry from which statistics will be sourced.
+   */
+  public void appendStats(PartitionStats entry) {
+    this.specId = Math.max(specId, entry.specId);
```

Comment:
Same comment as in `liveEntry`, would make it a precondition instead.

Reply:
Here this logic is needed for schema evolution

 If the previous partition only had c1 with the value c1 = 'foo' and the spec evolves to
 include both c1 and c2, the old partition will be interpreted as c1 = 'foo' and c2 = 'null'.
 Hence, if new inserts contain records with c1 = 'foo' and c2 = 'null', the spec id needs to be updated accordingly.

Reply:
We can't actually combine stats across spec IDs even if the coerced key happens to be the same. Partitions that belong to different specs aren't equivalent. For instance, we use `PartitionMap` for assigning deletes to ensure that not only the partition keys match but also specs. If we have `p1 = 10` in spec 1 and `p1 = 10 and p2 = null` in spec 2, these partitions are different and their stats can't be combined.

Does this make sense? If so, we may need to use `PartitionMap` instead of `StructLikeMap`.

Reply:
> Does this make sense? If so, we may need to use PartitionMap instead of StructLikeMap.

Not yet. 
My point of view is based on query planner. 
I had a table partitioned by `p1` with `p1 = 10` with 5 records. Now partition has evolved into `p1` and `p2`. 
After evolution, the new insert has 6 records for `p1 = 10` and `p2 = null`. 

So, when a query has a filter on `p1 = 10` and `p2 = null`, planner should get the stats of 11 (5+6) records from both old and new spec. So, why not combine them with the updated spec id?

Reply:
Even though I don't have strong opinion on keeping stats separate for each spec. (reasoned in just above comment). 

I have pushed a separated commit to handle the same. So, it is addressed.